### PR TITLE
add more geoip service urls

### PR DESCRIPTION
**Pull request by mariobassem**
## Related issues:

- #1887

### CODE REVIEW

File: pkg/geoip/geoip.go [hunk 1]
```diff
@@ -26,19 +30,26 @@ func Fetch() (Location, error) {
 		City:      "Unknown",
 	}
 
-	resp, err := http.Get("https://geoip.grid.tf")
-	if err != nil {
-		return l, err
-	}
-	defer resp.Body.Close()
+	for i := 0; i < len(geoipURLs); i++ {
```
Comment (xmonader) [inline pkg/geoip/geoip.go]: why not a for range?

File: pkg/geoip/geoip.go [hunk 2]
```diff
@@ -26,19 +30,26 @@ func Fetch() (Location, error) {
...
+		resp, err := http.Get(geoipURLs[i])
+		if err != nil {
+			log.Err(err).Msgf("failed to make http call to geoip service %s. retrying...", geoipURLs[i])
+			continue
+		}
+		defer resp.Body.Close()
```
Comment (xmonader) [inline pkg/geoip/geoip.go]: this won't be properly handled (you're calling the defer within a for-loop)

**Review by xmonader — COMMENTED** [review]
I was expecting that we would fire up multiple requests in the background against the 3 geoip services and as soon one of the returns a healthy result we can use it, but  @muhamadazmy can decide how to approach it properly

File: pkg/geoip/geoip.go [hunk 3]
```diff
@@ -22,34 +22,34 @@ type Location struct {
...
 
-		if err := json.NewDecoder(resp.Body).Decode(&l); err != nil {
-			log.Err(err).Msgf("failed to decode location data from %s", geoipURLs[i])
-			continue
-		}
+	return Location{}, errors.New("failed to fetch location information")
+}
 
-		return l, nil
+func getLocation(url string) (Location, error) {
```
Comment (xmonader) [inline pkg/geoip/geoip.go]: see how many Location{} ? can use a named parameter or even create it on the top of the function and reuse it

File: pkg/geoip/geoip.go [hunk 4]
```diff
@@ -22,34 +22,34 @@ type Location struct {
...
-		City:      "Unknown",
-	}
-
-	for i := 0; i < len(geoipURLs); i++ {
-		resp, err := http.Get(geoipURLs[i])
+	for _, url := range geoipURLs {
+		l, err := getLocation(url)
 		if err != nil {
-			log.Err(err).Msgf("failed to make http call to geoip service %s. retrying...", geoipURLs[i])
+			log.Err(err).Msg("failed to fetch location. retrying...")
```
Comment (xmonader) [inline pkg/geoip/geoip.go]: encode which broken service

File: pkg/geoip/geoip.go (hunk 3 above)
Comment (xmonader) [inline pkg/geoip/geoip.go]: can use a better name for url ( e.g geoIPService) because it's not apparent from the first sight what url that is

**Review by xmonader — DISMISSED** [review]

File: pkg/geoip/geoip.go [hunk 5]
```diff
@@ -18,6 +20,22 @@ type Location struct {
 
 // Fetch retrieves the location of the system calling this function
 func Fetch() (Location, error) {
+	geoipURLs := []string{"https://geoip.grid.tf/", "https://02.geoip.grid.tf/", "https://03.geoip.grid.tf/"}
```
Comment (muhamadazmy) [inline pkg/geoip/geoip.go]: can this be defined in the global scope (outside the function body) not on the stack. it's more of a constant than a variable. like

```go
var (
   geoipURLs = []string{...}
)
```

File: pkg/geoip/geoip.go [hunk 6]
```diff
@@ -18,6 +20,22 @@ type Location struct {
...
+
+	for _, url := range geoipURLs {
+		l, err := getLocation(url)
+		if err != nil {
+			log.Err(err).Msgf("failed to fetch location from geoip service %s. retrying...", url)
```
Comment (muhamadazmy) [inline pkg/geoip/geoip.go]: the log here is kinda misleading, we are not `retrying` instead we are trying a completely different url. Instead the error should show which url is not working + the cause. Also please use the structured logs correctly. Use the available, Str, Int, etc.. not formatting as much as possiblel

```go
log.Err(err).Str("url", url).Msg("failed to fetch location from geoip service")
```

Is more correct form than the one in code

**Review by muhamadazmy — CHANGES_REQUESTED** [review]

**Review by muhamadazmy — APPROVED** [review]
Please merge only after testing

**Pull request merged**